xbmc/cores/AudioEngine/Sinks/AESinkXAudio.cpp

```diff
@@ -887,3 +903,48 @@ bool CAESinkXAudio::IsUSBDevice()
 #endif
   return false;
 }
+
+bool CAESinkXAudio::AddEndOfStreamPacket()
+{
+  constexpr unsigned int frames{1};
+
+  XAUDIO2_BUFFER xbuffer = BuildXAudio2Buffer(nullptr, frames, 0);
+  xbuffer.Flags = XAUDIO2_END_OF_STREAM;
+
+  HRESULT hr = m_sourceVoice->SubmitSourceBuffer(&xbuffer);
+
+  if (hr != S_OK)
+  {
+    CLog::LogF(LOGERROR, "SubmitSourceBuffer failed due to {:X}", hr);
+    delete xbuffer.pContext;
+    return false;
+  }
+
+  m_sinkFrames += frames;
+  m_framesInBuffers += frames;
```

Comment:
As this additional frame is a frame AE is unaware of, we need to make sure it is nowhere used in reporting back to AE as this would make AE's error computation problematic. If that's not the case and only used internally to track Flush / blocking drain operation all good.

Reply:
Above comment is duplicate of the other comment.

Reply:
The queued amount of frames _after_ Drain should be zero. Same for Flush (Deinitialize), while one plays them out (drain) and the other asks to throw them away (Deintiialize).

AE calls them sometimes in order when a stream ends for example.

Reply:
yes it is zero in both cases because the sink now waits for the BufferEnd callbacks, which decrease the count, As a safety if the waits don't complete in a reasonable time, or if for some reason one of the callbacks doesn't happen, the count is reset to zero, at the same time as the sinkFrames.